contracts/exchangeIssuance/ExchangeIssuance.sol

```diff
@@ -0,0 +1,768 @@
+    */
+    function getAmountInToIssueExactSet(
+        ISetToken _setToken,
+        IERC20 _inputToken,
+        uint256 _amountSetToken
+    )
+        isSetToken(_setToken)
+        external
+        view
+        returns(uint256)
+    {
+        require(_amountSetToken > 0, "ExchangeIssuance: INVALID INPUTS");
+        
+        uint256 totalEth = 0;
+        
+        ISetToken.Position[] memory positions = _setToken.getPositions();
+        for(uint256 i = 0; i < positions.length; i++) {     
+            uint256 amountToken = uint256(positions[i].unit).mul(_amountSetToken).div(1 ether);
+            
+            // get minimum amount of ETH to be spent to acquire the required amount of tokens
+            (uint256 amountEth,) = _getMinTokenForExactToken(amountToken, WETH, positions[i].component);
+            totalEth = totalEth.add(amountEth);
+        }
+        
+        if(address(_inputToken) == WETH) {
+            return totalEth;
+        }
+        
+        (uint256 tokenAmount, ) = _getMinTokenForExactToken(totalEth, address(_inputToken), address(WETH));
+        return tokenAmount;
+    }
+    
+    /**
+     * Returns an estimated amount of ETH or specified ERC20 received for a given SetToken and SetToken amount. 
+     *
+     * @param _setToken             SetToken redeemed
+     * @param _amountSetToRedeem    Amount of SetToken
+     * @param _outputToken          Address of output token. Ignored if _isOutputETH is true
+     * @return                      Estimated amount of ether/erc20 that will be received
+     */
+    function getEstimatedRedeemSetAmount(
+        ISetToken _setToken,
+        address _outputToken,
+        uint256 _amountSetToRedeem
+    ) 
+        isSetToken(_setToken)
+        external
+        view
+        returns (uint256)
+    {
+        require(_amountSetToRedeem > 0, "ExchangeIssuance: INVALID INPUTS");
+        
+        ISetToken.Position[] memory positions = _setToken.getPositions();
+        uint256 totalEth = 0;
+        for (uint256 i = 0; i < positions.length; i++) {
+            uint256 amount = uint256(positions[i].unit).mul(_amountSetToRedeem).div(1 ether);
+            
+            // get maximum amount of ETH received for a given amount of SetToken component
+            (uint256 amountEth, ) = _getMaxTokenForExactToken(amount, positions[i].component, WETH);
+            totalEth = totalEth.add(amountEth);
+        }
+        if(_outputToken == WETH) {
+            return totalEth;
+        }
+        
+        // get maximum amount of tokens for totalEth amount of ETH
+        (uint256 tokenAmount, ) = _getMaxTokenForExactToken(totalEth, WETH, _outputToken);
+        return tokenAmount;
+    }
+    
+    
+    /* ============ Internal Functions ============ */
+
+    /**
+     * Sets a max aproval limit for an ERC20 token, provided the current allowance is zero. 
+     * 
+     * @param _token    Token to approve
+     * @param _spender  Spender address to approve
+     */
```

Comment:
Please add @return javadoc for all internal functions. We have had bugs historically due to lack of clear documentation of return values